src/error.rs

```diff
@@ -51,52 +50,38 @@ impl From<String> for ErrorResponse {
-            error: value.to_owned(),
-        }
-    }
-}
-
-impl IntoResponse for HcHttpGatewayError {
-    fn into_response(self) -> axum::response::Response {
+impl HcHttpGatewayError {
+    /// Convert error into HTTP status code and error message.
+    pub fn into_status_code_and_body(self) -> (StatusCode, String) {
         match self {
-            HcHttpGatewayError::RequestMalformed(e) => (
-                StatusCode::BAD_REQUEST,
-                Json(ErrorResponse::from(format!("Request is malformed: {e}"))),
-            ),
-            HcHttpGatewayError::UnauthorizedFunction {
-                app_id,
-                zome_name,
-                fn_name,
-            } => (
-                StatusCode::FORBIDDEN,
-                Json(ErrorResponse::from(format!(
-                    "Function {fn_name} in zome {zome_name} in app {app_id} is not allowed"
-                ))),
-            ),
+            HcHttpGatewayError::RequestMalformed(_) => (StatusCode::BAD_REQUEST, self.to_string()),
+            HcHttpGatewayError::UnauthorizedFunction { .. } => {
+                (StatusCode::FORBIDDEN, self.to_string())
+            }
             HcHttpGatewayError::UpstreamUnavailable => (
                 StatusCode::BAD_GATEWAY,
-                Json(ErrorResponse::from("Could not connect to Holochain")),
-            ),
-            HcHttpGatewayError::AppSelectionError(AppSelectionError::NotInstalled) => (
-                StatusCode::NOT_FOUND,
-                Json(ErrorResponse::from(self.to_string())),
-            ),
-            HcHttpGatewayError::AppSelectionError(AppSelectionError::NotAllowed) => (
-                StatusCode::FORBIDDEN,
-                Json(ErrorResponse::from(self.to_string())),
-            ),
-            HcHttpGatewayError::AppSelectionError(AppSelectionError::MultipleMatching) => (
-                StatusCode::INTERNAL_SERVER_ERROR,
-                Json(ErrorResponse::from(self.to_string())),
+                "Could not connect to Holochain".to_string(),
             ),
+            HcHttpGatewayError::AppSelectionError(AppSelectionError::NotInstalled) => {
+                (StatusCode::NOT_FOUND, self.to_string())
+            }
+            HcHttpGatewayError::AppSelectionError(AppSelectionError::NotAllowed) => {
+                (StatusCode::FORBIDDEN, self.to_string())
+            }
+            HcHttpGatewayError::AppSelectionError(AppSelectionError::MultipleMatching) => {
+                (StatusCode::INTERNAL_SERVER_ERROR, self.to_string())
+            }
             _ => (
                 StatusCode::INTERNAL_SERVER_ERROR,
-                Json(ErrorResponse::from("Something went wrong")),
+                "Something went wrong".to_string(),
             ),
         }
-        .into_response()
+    }
+}
+
+impl IntoResponse for HcHttpGatewayError {
+    fn into_response(self) -> axum::response::Response {
+        let (status_code, body) = self.into_status_code_and_body();
```

Comment:
I've been thinking since I first saw this module that there's got to be a way to make this less verbose. Happy now.